File: problems/problems_3202/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"


using namespace std;
using json = nlohmann::json;

class Solution {
public:
    int maximumLength(const vector<int>& nums, int k) {
      vector<vector<int>> dp(k, vector<int>(k));
			int ans = 0;
			for (int num: nums) {
				num %= k;
				for (int i = 0; i < k; ++i) {
					dp[num][i] = dp[i][num] + 1;
					ans = max(ans, dp[num][i]);
				}
			}
			return ans;
    }
};

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	Solution solution;
	vector<int> nums = json::parse(inputArray.at(0));
	int k = json::parse(inputArray.at(1));
	return solution.maximumLength(nums, k);
}

File: problems/problems_3202/solution.py
import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.maximumLength(*test_input)

    def maximumLength(self, nums: List[int], k: int) -> int:
        # dp = [[0] * k for _ in range(k)]
        # for x in nums:
        #     cur = x % k
        #     for y, length in enumerate(dp[cur]):
        #         dp[y][cur] = length + 1
        # return max(map(max, dp))

        ans = 0
        for val in range(k):
            dp = [0] * k
            for num in nums:
                num %= k
                dp[num] = dp[val - num] + 1
            ans = max(ans, max(dp))
        return ans
